# Mass and isotopes

This tutorial includes following fundamental operations related to molecular mass and isotopes.

- Standard molecular weight
- Relative molecular mass (Calculated exact mass)
- Monoisotopic mass
- Isotopic composition
- Simulate mass spectrum

## Loading packages

This tutorial uses `Plots.jl` to display simulated mass spectrum.

In [1]:
import Pkg
Pkg.activate("..")
using MolecularGraph
using Plots
gr()
Plots.GRBackend()

# Convenient function to display a pair of mol images
function displayimgpair(img1, img2)
    """<div>
        <div style="float:left">$(img1)</div>
        <div style="float:left">$(img2)</div>
    </div>"""
end
;

Activating environment at `~/Workspace/MolecularGraph.jl_notebook/Project.toml`


┌ Info: Precompiling Plots [91a5bcdd-55d7-5caf-9e0b-520d859cae80]
└ @ Base loading.jl:1273


## Molecular weight (Relative molecular mass)

Molecular weight (Relative molecular mass) can be calculated by `standardweight` method.

In [2]:
standardweight(smilestomol("CCO"))

(46.069, 0.004529999999995926)

- `standardweight` returns the tuple of standard weight and its uncertainty, respectively. The result shown above means the molecular weight of C2H5O is 46.069±0.005.
- If you dont need uncertainty, give Float64 type to the method. Then the value will be rounded at 2 digits after the decimal point.

In [3]:
standardweight(Float64, smilestomol("CCO"))

46.07

## Calculated exact mass and monoisotopic mass

Both `exactmass` and `monoisotopicmass` returns exact mass of the given molecule. The difference is that `exactmass` considers atomic mass specified in `Atom.mass` fields of atoms whereas `monoisotopicmass` always returns exact mass of the molecule consists of the most abundant isotopes in terrestrial sources.

In [4]:
etohd6 = smilestomol("[2H]C([2H])([2H])C([2H])([2H])O[2H]")

println("Molecular weight: ", standardweight(Float64, etohd6))
println("Monoisotopic: ", monoisotopicmass(Float64, etohd6))
println("Exact: ", exactmass(Float64, etohd6))

Molecular weight: 52.11
Monoisotopic: 46.041865
Exact: 52.079525


## Isotopic composition

`isotopiccomposition(atomsymbol, num)` returns total masses of the given number of atoms and their isotopic compositions.

Isotopes that have lower abundance than the `threshold` parameter will be filtered out (default 0.001 = 0.1%)

In [16]:
isotopiccomposition(:C, 100; threshold=0.01)

5-element Array{Tuple{Float64,Float64},1}:
 (1204.01341934028, 0.01829972319797687)
 (1203.01006450521, 0.06977096630391633)
 (1202.00670967014, 0.19747098047720194)
 (1201.00335483507, 0.36875136628763155)
 (1200.0, 0.3410365354703909)           

In [15]:
isotopiccomposition(:C, 1000; threshold=0.01)

15-element Array{Tuple{Float64,Float64},1}:
 (12018.06038703126, 0.011690519214538609)
 (12017.05703219619, 0.019792334300846457)
 (12016.05367736112, 0.0316151371989973)  
 (12015.05032252605, 0.04748134937359465) 
 (12014.04696769098, 0.06678541451097929) 
 (12013.04361285591, 0.08758641222522137) 
 (12012.04025802084, 0.1065534144511583)  
 (12011.03690318577, 0.11953540283428146) 
 (12010.0335483507, 0.12279996975578453)  
 (12009.03019351563, 0.11456944727429401) 
 (12008.02683868056, 0.09610452890364467) 
 (12007.02348384549, 0.0715861101895721)  
 (12006.02012901042, 0.04661058275253684) 
 (12005.01677417535, 0.02598703123363034) 
 (12004.01341934028, 0.012061765298138035)

`isotopiccomposition(mol)` returns isotopic composition of the molecule.

In [17]:
isotopiccomposition(smilestomol("CCO"))

3-element Array{Tuple{Float64,Float64},1}:
 (46.041864812949996, 0.9756627354527866) 
 (47.045219648019994, 0.02109374866472255)
 (48.04610980624, 0.0020049807108054704)  

## Simulate mass spectrum

According to the isotopic composition, simulated mass spectrum can be plotted by using `simulatemassspec` method and available plot libraries.
Follwing example uses Plots.jl to draw the simulated spectrum.

In [18]:
mol = smilestomol("c1cc(ccc1[C@H]([C@@H](CO)NC(=O)C(Cl)Cl)O)[N+](=O)[O-]")
structure = drawsvg(mol, 200, 200)
data = simulatemassspec(mol)
p = plot(data[:, 1], data[:, 2], size=(320, 240), leg=false, xlabel = "Mass", ylabel = "Intensity")
buf = IOBuffer()
show(buf, MIME("image/svg+xml"), p)
spectrum = String(take!(buf))
close(buf)
display("text/html", displayimgpair(structure, spectrum))

H 
 N + 
 HO 
 Cl 
 O 
 H 
 HO 
 Cl 
 O – 
 NH 
 O 
 
 
 <?xml version="1.0" encoding="utf-8"?>
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
 
<path clip-path="url(#clip8100)" d="M 0 0 M367.304 812.418 Q370.661 813.135 372.536 815.404 Q374.434 817.672 374.434 821.006 Q374.434 826.121 370.916 828.922 Q367.397 831.723 360.916 831.723 Q358.74 831.723 356.425 831.283 Q354.133 830.867 351.68 830.01 L351.68 825.496 Q353.624 826.631 355.939 827.209 Q358.254 827.788 360.777 827.788 Q365.175 827.788 367.466 826.052 Q369.781 824.316 369.781 821.006 Q369.781 817.95 367.628 816.237 Q365.499 814.501 361.679 814.501 L357.652 814.501 L357.652 810.658 L361.865 810.658 Q365.314 810.658 367.142 809.293 Q368.971 807.904 368.971 805.311 Q368.971 802.649 367.073 801.237 Q365.198 799.802 361.679 799.802 Q359.758 799.802 357.559 800.219 Q355.36 800.635 352.721 801.515 L352.721 797.348 Q355.383 796.608 357.698 796.237 Q360.036 795.867 362.096 795.867 Q367.42 795.867 370.522 798.297 Q373.624 800.705 373.624 804.825 Q373.624 807.695 371.98 809.686 Q370.337 811.654 367.304 812.418 Z" fill="#000000" fill-rule="evenodd" fill-opacity="1" /><path clip-path="url(#clip8100)" d="M 0 0 M383.531 827.117 L399.851 827.117 L399.851 831.052 L377.906 831.052 L377.906 827.117 Q380.568 824.362 385.152 819.732 Q389.758 815.08 390.939 813.737 Q393.184 811.214 394.064 809.478 Q394.966 807.719 394.966 806.029 Q394.966 803.274 393.022 801.538 Q391.101 799.802 387.999 799.802 Q385.8 799.802 383.346 800.566 Q380.915 801.33 378.138 802.881 L378.138 798.159 Q380.962 797.024 383.415 796.446 Q385.869 795.867 387.906 795.867 Q393.277 795.867 396.471 798.552 Q399.665 801.237 399.665 805.728 Q399.665 807.858 398.855 809.779 Q398.068 811.677 395.962 814.27 Q395.383 814.941 392.281 818.158 Q389.179 821.353 383.531 827.117 Z" fill="#000000" fill-rule="evenodd" fill-opacity="1" /><path clip-path="url(#clip8100)" d="M 0 0 M414.92 814.64 Q411.587 814.64 409.665 816.422 Q407.767 818.205 407.767 821.33 Q407.767 824.455 409.665 826.237 Q411.587 828.019 414.92 828.019 Q418.253 828.019 420.175 826.237 Q422.096 824.432 422.096 821.33 Q422.096 818.205 420.175 816.422 Q418.276 814.64 414.92 814.64 M410.244 812.649 Q407.235 811.908 405.545 809.848 Q403.878 807.788 403.878 804.825 Q403.878 800.682 406.818 798.274 Q409.781 795.867 414.92 795.867 Q420.082 795.867 423.022 798.274 Q425.962 800.682 425.962 804.825 Q425.962 807.788 424.272 809.848 Q422.605 811.908 419.619 812.649 Q422.999 813.436 424.874 815.728 Q426.772 818.02 426.772 821.33 Q426.772 826.353 423.693 829.038 Q420.637 831.723 414.92 831.723 Q409.202 831.723 406.124 829.038 Q403.068 826.353 403.068 821.33 Q403.068 818.02 404.966 815.728 Q406.864 813.436 410.244 812.649 M408.531 805.265 Q408.531 807.95 410.198 809.455 Q411.888 810.959 414.92 810.959 Q417.929 810.959 419.619 809.455 Q421.332 807.95 421.332 805.265 Q421.332 802.58 419.619 801.075 Q417.929 799.571 414.92 799.571 Q411.888 799.571 410.198 801.075 Q408.531 802.58 408.531 805.265 Z" fill="#000000" fill-rule="evenodd" fill-opacity="1" /><path clip-path="url(#clip8100)" d="M 0 0 M638.452 812.418 Q641.808 813.135 643.683 815.404 Q645.582 817.672 645.582 821.006 Q645.582 826.121 642.063 828.922 Q638.545 831.723 632.063 831.723 Q629.887 831.723 627.572 831.283 Q625.281 830.867 622.827 830.01 L622.827 825.496 Q624.771 826.631 627.086 827.209 Q629.401 827.788 631.924 827.788 Q636.322 827.788 638.614 826.052 Q640.929 824.316 640.929 821.006 Q640.929 817.95 638.776 816.237 Q636.646 814.501 632.827 814.501 L628.799 814.501 L628.799 810.658 L633.012 810.658 Q636.461 810.658 638.29 809.293 Q640.119 807.904 640.119 805.311 Q640.119 802.649 638.22 801.237 Q636.345 799.802 632.827 799.802 Q630.906 799.802 628.707 800.219 Q626.508 800.635 623.869 801.515 L623.869 797.348 Q626.531 796.608 628.846 796.237 Q631.183 795.867 633.244 795.867 Q638.568 795.867 641.67 798.297 Q644.771 800.705 644.771 804.825 Q644.771 807.695 643.128 809.686 Q641.484 811.654 638.452 812.418 Z" 